google/_modules/gke/node_pool/variables.tf

```diff
@@ -115,3 +115,13 @@ variable "node_locations" {
   description = "List of zones in the cluster's region to start worker nodes in. Defaults to cluster's node locations."
   default     = null
 }
+
+variable "guest_accelerator" {
+  type        = map(any)
```

Comment:
Could we use object here instead of map(any). Any should be the last resort in my opinion. Not the first thing we try.

https://developer.hashicorp.com/terraform/language/expressions/type-constraints#optional-object-type-attributes

Reply:
If we use `object({})` as type what happened is any user need to explicit calling all keys under that object:

```
  type = object({
    type               = string
    count              = number
    gpu_partition_size = string
    gpu_sharing_config = object({
      gpu_sharing_strategy       = string
      max_shared_clients_per_gpu = number
    })
  })
```

Then anyone want to use `guest_accelerator` they have to provide all keys which might be difficult for people who doesn't know about the params and how to use it at first. And we can't set default for nested stuff either 😢. 

```
guest_accelerator = {
  type = "A100"
  count = 1
  # even though they don't want to use it. They have to set to something that match the object({}) above.
  gpu_partition_size = null || ""
  gpu_sharing_config = {
    gpu_sharing_strategy             = null || ""
    max_shared_clients_per_gpu = null || 2 
    # and we may need additional check for `null` and `0` or `""`.
  }
}
```


While with the current approach they can gradually expand the options while researching more about upstream docs.

Another way I can think of is not using `map(any)` but add more variables to better do default and validations on top of it. So we can have vars with default value for usability.

```
# These 2 likely are important and often will be used
var "guest_accelerator_type" {}
var "guest_accelerator_count" {} # default to 0 which we can use this for the `dynamic` block to create it or not.

# Others will all default to null and with proper type.
var "guest_accelerator_gpu_partition_size" {} # default null
var "guest_accelerator_gpu_sharing_config" {} # default null, `object()` type.
...
```

WDYT?

Reply:
Check the link, you can mark the attributes optional.

Reply:
Where was I when this was announced 😂. Will fix it later.